web/packages/teleterm/src/services/config/configService.ts

```diff
@@ -14,168 +14,119 @@
-     * in a styled component or used in CSS, as it may inject malicious CSS code.
-     * Before using it, sanitize it with `CSS.escape` or pass it as a `style` prop.
-     * Read more https://frontarm.com/james-k-nelson/how-can-i-use-css-in-js-securely/.
-     */
-    'terminal.fontFamily': z.string().default(defaultTerminalFont),
-    'terminal.fontSize': z.number().int().min(1).max(256).default(15),
-  });
-};
-
-export type AppConfig = z.infer<ReturnType<typeof createAppConfigSchema>>;
-
-export type KeyboardShortcutAction =
-  | 'tab1'
-  | 'tab2'
-  | 'tab3'
-  | 'tab4'
-  | 'tab5'
-  | 'tab6'
-  | 'tab7'
-  | 'tab8'
-  | 'tab9'
-  | 'closeTab'
-  | 'newTab'
-  | 'previousTab'
-  | 'nextTab'
-  | 'openQuickInput'
-  | 'openConnections'
-  | 'openClusters'
-  | 'openProfiles';
-
-const getDefaultKeymap = (platform: Platform) => {
-  switch (platform) {
-    case 'win32':
-      return {
-        tab1: 'Ctrl+1',
-        tab2: 'Ctrl+2',
-        tab3: 'Ctrl+3',
-        tab4: 'Ctrl+4',
-        tab5: 'Ctrl+5',
-        tab6: 'Ctrl+6',
-        tab7: 'Ctrl+7',
-        tab8: 'Ctrl+8',
-        tab9: 'Ctrl+9',
-        closeTab: 'Ctrl+W',
-        newTab: 'Ctrl+T',
-        previousTab: 'Ctrl+Shift+Tab',
-        nextTab: 'Ctrl+Tab',
-        openQuickInput: 'Ctrl+K',
-        openConnections: 'Ctrl+P',
-        openClusters: 'Ctrl+E',
-        openProfiles: 'Ctrl+I',
-      };
-    case 'linux':
-      return {
-        tab1: 'Alt+1',
-        tab2: 'Alt+2',
-        tab3: 'Alt+3',
-        tab4: 'Alt+4',
-        tab5: 'Alt+5',
-        tab6: 'Alt+6',
-        tab7: 'Alt+7',
-        tab8: 'Alt+8',
-        tab9: 'Alt+9',
-        closeTab: 'Ctrl+W',
-        newTab: 'Ctrl+T',
-        previousTab: 'Ctrl+Shift+Tab',
-        nextTab: 'Ctrl+Tab',
-        openQuickInput: 'Ctrl+K',
-        openConnections: 'Ctrl+P',
-        openClusters: 'Ctrl+E',
-        openProfiles: 'Ctrl+I',
-      };
-    case 'darwin':
+import { createAppConfigSchema, AppConfig } from './createAppConfigSchema';
+
+const logger = new Logger('ConfigService');
+
+export interface ConfigService {
+  get<K extends keyof AppConfig>(
+    key: K
+  ): { value: AppConfig[K]; metadata: { isStored: boolean } };
+
+  set<K extends keyof AppConfig>(key: K, value: AppConfig[K]): void;
+
+  getStoredConfigErrors(): ZodIssue[] | undefined;
+}
+
+export function createConfigService({
+  configFile,
+  jsonSchemaFile,
+  platform,
+}: {
+  configFile: FileStorage;
+  jsonSchemaFile: FileStorage;
+  platform: Platform;
+}): ConfigService {
+  const schema = createAppConfigSchema(platform);
+  updateJsonSchema({ schema, configFile, jsonSchemaFile });
+
+  const { storedConfig, configWithDefaults, errors } = validateStoredConfig();
+
+  function parse(data: Partial<AppConfig>) {
+    return schema.safeParse(data);
+  }
+
+  //TODO (gzdunek): syntax errors of the JSON file are silently ignored, report
+  // them to the user too
+  function validateStoredConfig(): {
```

Comment:
Yeah, much cleaner now. 

>This would be a good counterexample of breaking functions into smaller ones and letting them operate on different levels of abstractions vs the first version of updateJSonSchema.

Agree.